MediaBrowser.Controller/Entities/AudioBooks/AudioBookFile.cs

```diff
@@ -0,0 +1,205 @@
+                    nextChapters.Add(chapter);
+                }
+            }
+
+            return nextChapters;
+        }
+
+        public override double GetDefaultPrimaryImageAspectRatio()
+        {
+            return 0;
+        }
+
+        /// <summary>
+        /// Get user data keys.
+        /// </summary>
+        /// <returns>User data keys associated with this item.</returns>
+        public override List<string> GetUserDataKeys()
+        {
+            var list = base.GetUserDataKeys();
+
+            return list;
+        }
+
+        public string FindSeriesPresentationUniqueKey()
+        {
+            return AudioBookEntity?.PresentationUniqueKey;
+        }
+
+        // TODO: Behavior may be more consistent if we use "Album" metadata and push up to parent
+        public string FindAudioBookName()
+        {
+            return AudioBookEntity is null ? "AudioBook Unknown" : AudioBookEntity.Name;
+        }
+
+        public Guid FindAudioBookId()
+        {
+            return AudioBookEntity is null ? Guid.Empty : AudioBookEntity.Id;
+        }
+
+        /// <summary>
+        /// Creates the string by which this item will be sorted amongst other children. Logically, AudioBookFiles
+        /// must already be named in a unique, indexable way so we can just return the name.
+        /// </summary>
+        /// <returns>System.String.</returns>
+        protected override string CreateSortName()
+        {
+            return Name;
+        }
+
+        public override IEnumerable<FileSystemMetadata> GetDeletePaths()
+        {
+            return new[]
+            {
+                new FileSystemMetadata
+                {
+                    FullName = Path,
+                    IsDirectory = IsFolder
+                }
+            }.Concat(GetLocalMetadataFilesToDelete());
+        }
+
+        // TODO: I'm still not sure where this fits into the larger framework.
+        public new AudioBookFileInfo GetLookupInfo()
+        {
+            var id = GetItemLookupInfo<AudioBookFileInfo>();
+
+            id.BookTitle = Album;
+            id.Artists = Artists;
+            id.Container = Container;
+            id.Chapter = (int)IndexNumber;
+
+            return id;
+        }
+
+        public override bool BeforeMetadataRefresh(bool replaceAllMetadata)
+        {
+            var hasChanges = base.BeforeMetadataRefresh(replaceAllMetadata);
+
+            return hasChanges;
+        }
+
+        public override List<ExternalUrl> GetRelatedUrls()
+        {
+            var list = base.GetRelatedUrls();
+
+            // TODO: Is there one of these for books and are those changes worth it?
+
+            return list;
+        }
+
+        protected override IEnumerable<(BaseItem Item, MediaSourceType MediaSourceType)> GetAllItemsForMediaSources()
+        {
+            return [(this, MediaSourceType.Default)];
+        }
+
+        public void SetFilesPlayed(User user, IUserDataManager userDataManager)
```

Comment:
I know this is how it's done in the past, but we generally don't like having this much logic in the entities as "untangling" them becomes a big hassle. Entities should contain some state and operations on their own values, but bringing in other dependencies is too much.

Reply:
Good point. Do you have any thoughts on where I should put this logic?